File: 198. House Robber.py
#Time Complexity:O(n)
#Space Complexity:O(1)
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        rob1,rob2=0,0
        for n in nums:
            maxi=max(rob1+n,rob2)
            rob1=rob2
            rob2=maxi
        return maxi

File: 198. House Robber.cpp
class Solution {
public:
    /*
    Time complexity: O(N)
    Space complexity: O(1)  (modifying the input array)
    */
    int rob(vector<int>& a) {
      int n = a.size();

      //we can actually use 'a' as the dp array

      for(int i=1;i<n;i++){
        //2 options (pick bigger of these 2): 
        //1. take the maximum till last index (= dp[i-1]) (can't add a[i] since it will be taking adjacent cell)
        //2. take the maximum till last's previous (= dp[i-2]) and add current value a[i]
        if(i>1)  a[i] = max(a[i-2]+a[i], a[i-1]);
        else  a[i] = max(a[i], a[i-1]);
      }

      return a[n-1];
    }
};
